File: src/DawnlightSearch/QueryWorker/query_thread.py
from __future__ import absolute_import

from .._Global_Qt_import import *
from .._Global_DawnlightSearch import *
from .._Global_logger import *
from .sql_formatter import format_sql_cmd

try:
    import Queue
except:
    import queue as Queue

class QueryThread(QtCore.QThread):
    add_row_to_model_SIGNAL = QtCore.pyqtSignal(int, list)
    update_progress_SIGNAL = QtCore.pyqtSignal(int, int)

    def __init__(self, mutex, queue_condition, qqueue, parent=None,thread_id=0):
        super(self.__class__, self).__init__()
        self.mutex = mutex
        self.queue_condition = queue_condition
        self.qqueue = qqueue
        self.quit_flag = False
        self.thread_id = thread_id
        # self.add_row_to_model_SIGNAL = QtCore.pyqtSignal(list)

        # con = QtSql.QSqlDatabase.addDatabase("QSQLITE")
        # con.setDatabaseName(DATABASE_FILE_NAME)
        # if not con.open():
        #     print con.lastError().text()
        con = sqlite3.connect(DATABASE_FILE_NAME, check_same_thread=False)
        
        # note:  "COLLATE nocase" has no effect here. Use "(?i)" in the reg expr.
        import re
        def sqlite_regexp(expr, text):
            reg = re.compile(expr)
            return reg.search(text) is not None
        con.create_function("regexp", 2, sqlite_regexp)
        
        cur = con.cursor()

        self.cur = cur
        self.con = con

    def quit_previous_query(self):
        pass
        # TODO: cancel query

    def run(self):
        cur = self.cur
        cur.execute("PRAGMA case_sensitive_like=ON;")

        while (1):
            if self.quit_flag:
                break
            self.mutex.lock()
            self.queue_condition.wait(self.mutex, 1000 * 50)
            # print 'Condition wait:' ,QtCore.QThread.currentThreadId()
            # logger.debug('Condition wait')
            self.mutex.unlock()

            while not self.qqueue.empty():
                # print "Queue not empty: ", QtCore.QThread.currentThreadId()
                # logger.debug('Queue not empty')
                if self.quit_flag:
                    break

                self.mutex.lock()
                if self.qqueue.empty():
                    self.mutex.unlock()
                    continue  # break
                q = self.qqueue.get()
                self.mutex.unlock()


                query_id = q['query_id']
                sql_comm = q['sql_comm']
                # msg = "Queue get: " + str(q) + '\n\t\t' + "Query_Text_ID: " + str(GlobalVar.Query_Text_ID) + " ID: " + str(query_id)
                # logger.debug(msg)
                case_sensitive_like_flag_ON = q['case_sensitive_like_flag']
                if (query_id < GlobalVar.Query_Text_ID or GlobalVar.CURRENT_MODEL_ITEMS>=GlobalVar.MODEL_MAX_ITEMS):
                    continue
                try:
                    if case_sensitive_like_flag_ON:
                        cur.execute('PRAGMA case_sensitive_like=ON;')
                    else:
                        cur.execute('PRAGMA case_sensitive_like=OFF;')
                    cur.execute(sql_comm)
                    for query_row in cur:
                        if self.quit_flag:
                            break
                        row = [None] * len(DB_HEADER_LIST)
                        for idx, col in enumerate(query_row):
                            newitem = QtGui.QStandardItem(str(col))
                            newitem.setData(str(col), HACKED_QT_EDITROLE)
                            if idx in [QUERY_HEADER.Filename, QUERY_HEADER.Path]:
                                newitem.setData(str(col), HACKED_QT_EDITROLE)
                            if idx in [QUERY_HEADER.Size]:
                                newitem.setData(QtCore.QVariant(col), HACKED_QT_EDITROLE)
                            if idx in [QUERY_HEADER.atime, QUERY_HEADER.ctime, QUERY_HEADER.mtime]:
                                newitem.setData(QtCore.QVariant(col), HACKED_QT_EDITROLE)
                            if idx in [QUERY_HEADER.IsFolder]:
                                newitem.setTextAlignment(QtCore.Qt.AlignVCenter | QtCore.Qt.AlignHCenter)
                                if str(col) == '1':
                                    row[DB_HEADER.Extension] = QtGui.QStandardItem('')

                            row[QUERY_TO_DSP_MAP[idx]] = newitem

                            if idx == QUERY_HEADER.Filename and (row[DB_HEADER.IsFolder] is None):
                                extension = os.path.splitext(col)[1].replace('.','')
                                newitem = QtGui.QStandardItem(extension)
                                newitem.setData(extension, HACKED_QT_EDITROLE)
                                newitem.setTextAlignment(QtCore.Qt.AlignVCenter)
                                row[DB_HEADER.Extension] = newitem
                        if (query_id < GlobalVar.Query_Text_ID or GlobalVar.CURRENT_MODEL_ITEMS >= GlobalVar.MODEL_MAX_ITEMS):
                            break

                        self.add_row_to_model_SIGNAL.emit(query_id, row)
                        # self.model.appendRow(row)
                except Exception as e:
                    logger.error(str(e))
                if (not self.quit_flag) and (query_id == GlobalVar.Query_Text_ID):
                    self.update_progress_SIGNAL.emit(self.qqueue.qsize(), q['LEN'])

                    if self.qqueue.qsize() == 0:
                        GlobalVar.QUERY_READY_FLAG[self.thread_id] = True
                        # self.mutex.lock()
                        if   all(flag for flag in GlobalVar.QUERY_READY_FLAG):
                            # print(GlobalVar.QUERY_READY_FLAG)
                            logger.info('Query END')
                            self.update_progress_SIGNAL.emit(0, q['LEN'])
                        # self.mutex.unlock()

    def quit(self):
        self.quit_flag = True
        super(self.__class__, self).quit()

class DistributeQueryWorker(QtCore.QThread):
    def __init__(self, parent, **kwargs):  # *args, **kwargs
        super(self.__class__, self).__init__(parent)  # , **kwargs
        self.mutex = QtCore.QMutex()
        self.condition = QtCore.QWaitCondition()

        self.queue_mutex = QtCore.QMutex()
        self.queue_condition = QtCore.QWaitCondition()
        self.quit_flag = False

        self.qqueue = Queue.Queue()
        # In python, we don't need mutex. Queue is thread-safe.
        self.target_slot = kwargs['target_slot']
        self.update_progress_slot = kwargs['progress_slot']
        self.thread_no = max(1, QtCore.QThread.idealThreadCount())
        self.thread_pool = [QueryThread(self.queue_mutex, self.queue_condition, self.qqueue, parent=self, thread_id = i ) \
                            for i in range(self.thread_no)]
        GlobalVar.QUERY_READY_FLAG = [False] * self.thread_no
        for thread in self.thread_pool:
            thread.add_row_to_model_SIGNAL.connect(self.target_slot)
            thread.update_progress_SIGNAL.connect(self.update_progress_slot)
            thread.start()

    def run(self):
        while 1:
            if self.quit_flag:
                break
            self.mutex.lock()
            self.condition.wait(self.mutex)
            self.mutex.unlock()
            if self.quit_flag:
                break

            self.mutex.lock()
            [query_id, uuid_path_list, sql_text] = self.new_query_list
            self.mutex.unlock()

            # # ======================================
            # header_list = DB_HEADER_LIST
            # sql_mask = " SELECT " + ",".join(header_list) + ' FROM `%s` '
            # sql_mask = sql_mask.replace("Path", '''"%s"||Path''')
            # sql_where = []
            # for i in sql_text.split(' '):
            #     if i:
            #         sql_where.append(''' (`Filename` LIKE "%%%%%s%%%%") ''' % i)  # FIXME: ugly sql cmd
            # sql_where = " WHERE " + " AND ".join(sql_where)
            # sql_mask = sql_mask + sql_where
            # logger.debug("SQL mask: " + sql_mask)
            # # ======================================
            # logger.info(
            #     'distribute_new_query slot received: ' + str([query_id, uuid_path_list, sql_mask, cur_query_id_list
            #                                                   ]))
            # for thread in self.thread_pool:
            #     thread.quit_previous_query()
            #
            # self.queue_mutex.lock()
            # while not self.qqueue.empty():
            #     self.qqueue.get()
            # tmp_q = []
            # for table in uuid_path_list:  # {'uuid':uuid,'path':path,'rows':rows}
            #     table['path'] = "" if table['path'] == "/" else table['path']  # avoid "//dir/file"
            #     sql_comm = sql_mask % (table['path'], table['uuid'])  # path, uuid
            #     rowid_low = 0
            #     while rowid_low <= table['rows']:
            #         rowid_high = rowid_low + GlobalVar.QUERY_CHUNK_SIZE
            #         if "WHERE" in sql_comm.upper():
            #             sql_comm_2 = sql_comm + '  AND  (ROWID BETWEEN %s AND %s) ' % (rowid_low, rowid_high)
            #         else:
            #             sql_comm_2 = sql_comm + ' WHERE (ROWID BETWEEN %s AND %s) ' % (rowid_low, rowid_high)
            #         sql_comm_2 += " LIMIT %d" % GlobalVar.QUERY_LIMIT
            #         tmp_q.append({'query_id': query_id, 'sql_comm': sql_comm_2})
            #         rowid_low = rowid_high + 1
            # for idx, item in enumerate(tmp_q):
            #     item['SN'] = idx
            #     item['LEN'] = len(tmp_q)
            #     self.qqueue.put(item)
            # # ======================



            logger.info(
                'distribute_new_query slot received: ' + str([query_id, uuid_path_list, sql_text  ]))
            for thread in self.thread_pool:
                thread.quit_previous_query()

            self.queue_mutex.lock()
            while not self.qqueue.empty():
                self.qqueue.get()
            tmp_q = []
            for table in uuid_path_list:  # {'uuid':uuid,'path':path,'rows':rows}
                if not table['path']:       # table['path'] == None
                    # table['path'] = table['uuid']+'::'
                    table['path'] = table['alias'] + '::'
                table['path'] = "" if table['path'] == "/" else table['path']  # avoid "//dir/file"

                (table['path'], table['uuid'])  # path, uuid

                rowid_low = 0
                while rowid_low <= table['rows']:
                    rowid_high = rowid_low + GlobalVar.QUERY_CHUNK_SIZE
                    OK_flag, _ ,_ , sql_cmd, case_sensitive_like_flag_ON,_ = format_sql_cmd(
                        {
                            'path': table['path'],
                            'uuid': table['uuid'],
                            'sql_text': sql_text,
                            'rowid_low': rowid_low,
                            'rowid_high': rowid_high,
                        }
                    )
                    if OK_flag:
                        tmp_q.append({'query_id': query_id, 'sql_comm': sql_cmd, 'case_sensitive_like_flag': case_sensitive_like_flag_ON})
                    rowid_low = rowid_high + 1
            for idx, item in enumerate(tmp_q):
                item['SN'] = idx
                item['LEN'] = len(tmp_q)
                self.qqueue.put(item)

            GlobalVar.QUERY_READY_FLAG = [False] * self.thread_no

            self.queue_condition.wakeAll()

            self.queue_mutex.unlock()

    @pyqtSlot(list)
    def distribute_new_query(self, new_query_list):

        tmp_mutexlocker = QtCore.QMutexLocker(self.mutex)
        self.new_query_list = new_query_list
        self.condition.wakeOne()

    def quit(self):
        self.quit_flag = True
        for thread in self.thread_pool:
            thread.quit_flag = True
            thread.quit_previous_query()
            thread.quit()

        self.queue_mutex.lock()
        self.queue_condition.wakeAll()
        self.queue_mutex.unlock()

        self.mutex.lock()
        self.condition.wakeOne()
        self.mutex.unlock()

        for thread in self.thread_pool:
            thread.wait(5000)

        super(self.__class__, self).quit()

File: QueryWorker/query_thread.cpp
#include "query_thread.h"
int QueryThread::ERROR_QUERY_ID=-1;
QueryThread::QueryThread(QMutex * mutex_in, QWaitCondition * condition_in,
                         QString db_connection_name_in, QQueue<QPair<QPair<int,bool>, QString>> * sql_queue_in,
                         QObject *)
{
    mutex = mutex_in;
    condition = condition_in;
    db_connection_name = db_connection_name_in;
    sql_queue = sql_queue_in;

    restart = false;
    abort = false;
}

QueryThread::~QueryThread()
{
    //    mutex.lock();
    abort = true;
    //    condition.wakeOne();
    //    mutex.unlock();
    wait();
}



void QueryThread::run()
{
    bool ok;
    QSqlDatabase database;
    database = QSqlDatabase::addDatabase("QSQLITE", db_connection_name);
    database.setDatabaseName(DATABASE_FILE_NAME);
    if (! (database.open()))
    {
        qDebug()<< ("Fail to open db: "  +db_connection_name);
        return;
    }
    else
        qDebug()<<( "Open db OK: "  +db_connection_name);
    QSqlQuery cur(database);
    cur.setForwardOnly(true);

    // TODO: pragma for dbs other than sqlite
    cur.exec("PRAGMA case_sensitive_like=ON;");

    QPair<QPair<int, bool>, QString> pair;
    // open db
    forever {
        mutex->lock();
        while(sql_queue->isEmpty())
        {
            condition->wait(mutex);
            if (abort) break;
        }
        if (! sql_queue->isEmpty())
            pair = sql_queue->dequeue();
        else
            pair.second = "";

        mutex->unlock();

        if(abort)
            break;
        if (pair.second == "")
            continue;
        // QPair<int,QString> pair
        int query_id = pair.first.first;
        bool case_sensitive_like_flag_ON = pair.first.second;
        QString sql = pair.second;

        if (query_id==Query_Text_ID && query_id>ERROR_QUERY_ID )
        {
            // TODO: other than sqlite
            if (case_sensitive_like_flag_ON)
                ok = cur.exec("PRAGMA case_sensitive_like=ON;");
            else
                ok = cur.exec("PRAGMA case_sensitive_like=OFF;");
            if(!ok)
            {
                qDebug()<<"SQL ERROR: "<<__FILE__<< __LINE__<< __FUNCTION__<<
                          "\n\twhen: "<<cur.lastQuery()<<"\n\t"<<cur.lastError();
            }

            ok = cur.exec(sql);
            if (!ok) {
                ERROR_QUERY_ID = query_id;
                qDebug()<<"SQL ERROR: "<<__FILE__<< __LINE__<< __FUNCTION__<<
                          "\n\twhen: "<<cur.lastQuery()<<"\n\t"<<cur.lastError();
            }else
            {
                // return result
                // self.add_row_to_model_SIGNAL.emit(query_id, row)
                // add_row_to_model_SIGNAL(int query_id, QList<QList<QStandardItem *> > list_of_row_to_insert)

                if (query_id==Query_Text_ID &&  CURRENT_MODEL_ITEMS<MODEL_MAX_ITEMS)
                {
                    // build rows_of_item
                    QList<QList<QVariant> > list_of_row_to_insert;

                    while(cur.next())
                    {
                        if (query_id!=Query_Text_ID )
                            break;
                        /*SELECT
                         * Filename,"/devpath/"||Path,Size,IsFolder,atime,mtime,ctime FROM `UUID`
                         * WHERE  (ROWID BETWEEN 1 AND 2) AND ((ctime<1453527153))  LIMIT 10  */
                        QList<QVariant> row_to_insert;

                        for(int idx=0; idx< QUERY_HEADER.len;idx++){
                            row_to_insert  << cur.value(idx);

                        }// END per row loop
                        //                        qDebug()<<"emit row: "<<row_to_insert[0]->data(HACKED_QT_EDITROLE).toString()
                        //                                  <<row_to_insert[1]->data(HACKED_QT_EDITROLE).toString();

                          list_of_row_to_insert<<row_to_insert;
                    }// END whole list loop
                    if (list_of_row_to_insert.length()>0)
                    {
                        if (query_id!= Query_Text_ID)
                        {
                        }
                        else
                        {
//                            qDebug()<<"emit list";
                            emit  add_row_to_model_by_qvariant_SIGNAL( query_id, list_of_row_to_insert);

                        }
                    }
                }
            }
        }

        if(abort)
            break;

    }

    // close db
    database.close();
}

//===================================================================================

// ?? create item in this thread may sometime crash the main thread, when access data(Role)
void QueryThread::run_old_2()
{
    bool ok;
    QSqlDatabase database;
    database = QSqlDatabase::addDatabase("QSQLITE", db_connection_name);
    database.setDatabaseName(DATABASE_FILE_NAME);
    if (! (database.open()))
    {
        qDebug()<< ("Fail to open db: "  +db_connection_name);
        return;
    }
    else
        qDebug()<<( "Open db OK: "  +db_connection_name);
    QSqlQuery cur(database);
    cur.setForwardOnly(true);

    // TODO: pragma for dbs other than sqlite
    cur.exec("PRAGMA case_sensitive_like=ON;");

    QPair<QPair<int, bool>, QString> pair;
    // open db
    forever {
        mutex->lock();
        while(sql_queue->isEmpty())
        {
            condition->wait(mutex);
            if (abort) break;
        }
        if (! sql_queue->isEmpty())
            pair = sql_queue->dequeue();
        else
            pair.second = "";

        mutex->unlock();

        if(abort)
            break;
        if (pair.second == "")
            continue;
        // QPair<int,QString> pair
        int query_id = pair.first.first;
        bool case_sensitive_like_flag_ON = pair.first.second;
        QString sql = pair.second;

        if (query_id==Query_Text_ID && query_id>ERROR_QUERY_ID )
        {
            // TODO: other than sqlite
            if (case_sensitive_like_flag_ON)
                ok = cur.exec("PRAGMA case_sensitive_like=ON;");
            else
                ok = cur.exec("PRAGMA case_sensitive_like=OFF;");
            if(!ok)
            {
                qDebug()<<"SQL ERROR: "<<__FILE__<< __LINE__<< __FUNCTION__<<
                          "\n\twhen: "<<cur.lastQuery()<<"\n\t"<<cur.lastError();
            }

            ok = cur.exec(sql);
            if (!ok) {
                ERROR_QUERY_ID = query_id;
                qDebug()<<"SQL ERROR: "<<__FILE__<< __LINE__<< __FUNCTION__<<
                          "\n\twhen: "<<cur.lastQuery()<<"\n\t"<<cur.lastError();
            }else
            {
                // return result
                // self.add_row_to_model_SIGNAL.emit(query_id, row)
                // add_row_to_model_SIGNAL(int query_id, QList<QList<QStandardItem *> > list_of_row_to_insert)

                if (query_id==Query_Text_ID &&  CURRENT_MODEL_ITEMS<MODEL_MAX_ITEMS)
                {
                    // build rows_of_item
                    QList<QList<QStandardItem *> > list_of_row_to_insert;

                    QStandardItem * newitem;
                    QStandardItem * extension_item;
                    while(cur.next())
                    {
                        if (query_id!=Query_Text_ID )
                            break;
                        /*SELECT
                         * Filename,"/devpath/"||Path,Size,IsFolder,atime,mtime,ctime FROM `UUID`
                         * WHERE  (ROWID BETWEEN 1 AND 2) AND ((ctime<1453527153))  LIMIT 10  */
                        QList<QStandardItem *> row_to_insert;

                        for(int idx=0; idx< QUERY_HEADER.len;idx++){
                            //newitem = new QStandardItem(QString::number(idx));
                            newitem = new QStandardItem();
                            QVariant current_value = cur.value(idx);
                            if (current_value.isNull() || (!current_value.isValid()))
                                current_value = QVariant(1);
                            //                            newitem->setData(cur.value(idx), Qt::DisplayRole);
                            //                            newitem->setData(cur.value(idx), HACKED_QT_EDITROLE);
                            switch (idx) {
                            case QUERY_HEADER_class::Filename_:
                            {
                                QString pathname = current_value.toString();
                                newitem->setData(pathname, Qt::DisplayRole);
                                newitem->setData(pathname, HACKED_QT_EDITROLE);
                                row_to_insert << newitem;}
                                break;
                            case QUERY_HEADER_class::Path_:
                                //                                newitem->setData(cur.value(idx), HACKED_QT_EDITROLE);
                            {QString pathname = current_value.toString();
                                if(pathname.mid(0,2)=="//")  // modify path  like    //usr/bin
                                    pathname = pathname.mid(1);
                                newitem->setData(pathname, Qt::DisplayRole);
                                newitem->setData(pathname, HACKED_QT_EDITROLE);
                                row_to_insert << newitem;}
                                break;
                            case QUERY_HEADER_class::Size_:
                            case QUERY_HEADER_class::atime_:
                            case QUERY_HEADER_class::ctime_:
                            case QUERY_HEADER_class::mtime_:
                                //                                newitem->setData(cur.value(idx), HACKED_QT_EDITROLE);
                                newitem->setData(current_value, Qt::DisplayRole);
                                newitem->setData(current_value, HACKED_QT_EDITROLE);
                                row_to_insert << newitem;
                                break;
                            case QUERY_HEADER_class::IsFolder_:
                                // Alignment is controled by html_delegate, no effect here
//                                newitem->setTextAlignment(Qt::AlignHCenter|Qt::AlignVCenter);
                                newitem->setData(current_value.toBool(), Qt::DisplayRole);
                                newitem->setData(current_value.toBool(), HACKED_QT_EDITROLE);
                                //                                newitem->setData(cur.value(idx), HACKED_QT_EDITROLE);
                                row_to_insert << newitem;
                                extension_item =new QStandardItem();
                                if (current_value.toBool())
                                {
                                }
                                else{
                                    QString extension = cur.value(QUERY_HEADER_class::Filename_).toString();
                                    if (extension.contains("."))
                                        extension = extension.split(".").last();
                                    else
                                        extension = "";
                                    extension_item->setData(extension, HACKED_QT_EDITROLE);
                                    extension_item->setData(extension, Qt::DisplayRole);
                                }
                                row_to_insert << extension_item;
                                break;
                            default:
                                Q_ASSERT(false);
                                break;
                            }
                            QVariant test =  newitem->data(Qt::DisplayRole);
                            Q_ASSERT((row_to_insert.last() != nullptr));
                        }// END per row loop
                        //                        qDebug()<<"emit row: "<<row_to_insert[0]->data(HACKED_QT_EDITROLE).toString()
                        //                                  <<row_to_insert[1]->data(HACKED_QT_EDITROLE).toString();


                        for(int col=0; col < row_to_insert.length(); col++)
                        {

                            // FIXME
                            QVariant aaa = row_to_insert[col]->data(Qt::DisplayRole);
                        }


                        list_of_row_to_insert<<row_to_insert;
                    }// END whole list loop
                    if (list_of_row_to_insert.length()>0)
                    {
                        if (query_id!= Query_Text_ID)
                        {
                            // free memory
                            for(const QList<QStandardItem *>& list: list_of_row_to_insert)
                            {
                                for(QStandardItem * item: list)
                                    delete item;
                            }
                        }
                        else
                            emit  add_row_to_model_SIGNAL( query_id, list_of_row_to_insert);
                    }
                }
            }
        }

        if(abort)
            break;

    }

    // close db
    database.close();
}
